mlflow/types/utils.py

```diff
@@ -17,45 +17,71 @@ def __init__(self, msg):
         )
 
 
+def _get_tensor_shape(data: np.ndarray, variable_dimension: Optional[int] = 0) -> tuple:
+    """
+    Infer the shape of the inputted data.
+
+    This method creates the shape of the tensor to store in the TensorSpec. The variable dimension
+    is assumed to be the first dimension by default. This assumption can be overridden by inputting
+    a different variable dimension or `None` to represent that the input tensor does not contain a
+    variable dimension.
+
+    :param data: Dataset to infer from.
+    :param variable_dimension: An optional integer representing a variable dimension.
+    :return: tuple : Shape of the inputted data (including a variable dimension)
+    """
+    if not isinstance(data, np.ndarray):
+        raise TypeError("Expected numpy.ndarray, got '{}'.".format(type(data)))
+    variable_input_data_shape = data.shape
+    if variable_dimension is not None:
+        try:
+            variable_input_data_shape = list(variable_input_data_shape)
+            variable_input_data_shape[variable_dimension] = -1
+        except IndexError:
+            raise MlflowException(
+                "The specified variable_dimension {0} is out of bounds with"
+                "respect to the number of dimensions {1} in the input dataset".format(
+                    variable_dimension, data.ndim
+                )
+            )
+    return tuple(variable_input_data_shape)
+
+
 def _infer_schema(data: Any) -> Schema:
     """
     Infer an MLflow schema from a dataset.
 
-    This method captures the column names and data types from the user data. The signature
-    represents model input and output as data frames with (optionally) named columns and data
-    type specified as one of types defined in :py:class:`DataType`. This method will raise
-    an exception if the user data contains incompatible types or is not passed in one of the
-    supported formats (containers).
+    Data inputted as a numpy array or a dictionary is represented by :py:class:`TensorSpec`.
+    All other inputted data types are specified by :py:class:`ColSpec`.
+
+    A `TensorSpec` captures the data shape (default variable axis is 0), the data type (numpy.dtype)
+    and an optional name to represent the dataset.
```

Comment:
nit: I think we should say `an optional name for each individual tensor of the dataset` to mimic what we say for ColSpec.